openfl-workspace/torch_cnn_mnist/README.md

```diff
@@ -0,0 +1,212 @@
+    └── cnn_model.py      # centralized CNN model, ready for use in federated learning
+    ├── dataloader.py     # data loader module
+    └── taskrunner.py     # task runner module
+```
+
+## Directory Breakdown:
+* requirements.txt: This file lists all the Python dependencies required to run the TaskRunner API and its components. Ensure you install these dependencies by running pip install -r requirements.txt.
+* plan: Contains configuration files for federated learning:
+    - plan.yaml: The main Federated Learning plan declaration, defining the structure of the federated learning workflow.
+    - cols.yaml: A list of authorized collaborators for the federated learning task.
+    - data.yaml: Specifies the path to the data set for each collaborator.
+    - defaults: Path to the default configuration values for the federated learning plan.
+* src: This directory contains the Python modules used for federated learning:
+    - init.py: Marks the src directory as a Python package, allowing you to import modules within the directory.
+    - cnn_model.py: Defines the Convolutional Neural Network (CNN) model for federated learning.
+    - dataloader.py: A module responsible for loading and processing datasets for the federated learning task.
+    - taskrunner.py: The core task runner module that manages the execution of federated learning tasks.
+
+## Defining the Data Loader
+The data loader in OpenFL is responsible for batching and iterating through the dataset that will be used for local training and validation on each collaborator node. The PyTorchMNISTInMemory class is designed to handle the MNIST dataset, ensuring it is properly loaded and preprocessed for federated learning experiments.
+
+To customize the PyTorchMNISTInMemory class, you need to implement the load_mnist_shard() function to process the dataset available at data_path on the local file system. The data_path parameter represents the data shard number used by the collaborator. This setup allows each collaborator to work with a specific subset of the data, facilitating distributed training.
+
+The load_mnist_shard() function is responsible for loading the MNIST dataset, dividing it into training and validation sets, and applying necessary transformations. The data is then batched and made ready for the training process.
+
+# Modify the dataloader to support "bring your own data"
+You can either try to implement the placeholders by yourself, or get the solution from [dataloader.py](https://github.com/securefederatedai/openfl-contrib/blob/main/openfl_contrib_tutorials/ml_to_fl/federated/src/dataloader.py)
+
+```
+import numpy as np
+from typing import Iterator, Tuple
+from openfl.federated import PyTorchTaskRunner
+from openfl.utilities import Metric
+import torch.optim as optim
+import torch.nn.functional as F
+from src.cnn_model import DigitRecognizerCNN, train_epoch, validate
+
+class TemplateDataLoader(PyTorchDataLoader):
```

Comment:
I suggest renaming this class to `MNISTShardDataLoader`

Reply:
resolved in be37d1c1e35fee2cce4ee8907547c861e28e6d8f